### PR TITLE
[DataPipe] refactor to separate `_IterDataPipeMeta`

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70216
* #70373
* #76999
* #75995
* #70479
* __->__ #76776

There are some features of the metaclass that we would like to be unique to `IterDataPipe` and not applied to `MapDataPipe`. 

For example, `hook_iterator` should only impact `__iter__` of `IterDataPipe` since `MapDataPipe` has random access to the data and should not be constrained.

**Note that I plan to land this together the single iterator constraint PRs, so this will not land for now.**

Differential Revision: [D36156459](https://our.internmc.facebook.com/intern/diff/D36156459)

### CODE REVIEW

**Review by ejguan — APPROVED** [review]
LGTM, thanks a lot!

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36156459).

File: torch/utils/data/datapipes/_typing.py [hunk 1]
```diff
@@ -253,11 +253,9 @@ class _DataPipeMeta(GenericMeta):
     type: _DataPipeType
 
     def __new__(cls, name, bases, namespace, **kwargs):
-        if '__iter__' in namespace:
-            hook_iterator(namespace, 'enumerate(DataPipe)#{}'.format(name))
-
         return super().__new__(cls, name, bases, namespace, **kwargs)  # type: ignore[call-overload]
 
+        # TODO: the statements below are not reachable by design as there is a bug and typing is low priority for now.
```
Comment (ejguan) [inline torch/utils/data/datapipes/_typing.py:258]: Thanks for the comment for clarification!

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36156459).

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36156459).

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by VitalyFedyunin** [comment]
Suggestion, start landing this one now.

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36156459).

**Pull request closed without merging**